pimd/pim_cmd_common.c

```diff
@@ -653,3 +663,880 @@ int pim_process_no_rp_plist_cmd(struct vty *vty, const char *rp_str,
+			"\nActive Source           Group            RPT  IIF               OIL\n");
+	}
+
+	frr_each (rb_pim_oil, &pim->channel_oil_head, c_oil) {
+		char src_str[PIM_ADDRSTRLEN];
+		char grp_str[PIM_ADDRSTRLEN];
+		char in_ifname[INTERFACE_NAMSIZ + 1];
+		char out_ifname[INTERFACE_NAMSIZ + 1];
+		int oif_vif_index;
+		struct interface *ifp_in;
+		bool isRpt;
+
+		first_oif = 1;
+
+#if PIM_IPV == 4
+		if ((c_oil->up &&
+		     PIM_UPSTREAM_FLAG_TEST_USE_RPT(c_oil->up->flags)) ||
+		    c_oil->oil.mfcc_origin.s_addr == INADDR_ANY)
+			isRpt = true;
+#else
+		if ((c_oil->up &&
+		     PIM_UPSTREAM_FLAG_TEST_USE_RPT(c_oil->up->flags)) ||
+		    pim_addr_is_any(c_oil->oil.mf6cc_origin.sin6_addr))
+			isRpt = true;
+#endif
+		else
+			isRpt = false;
+
+#if PIM_IPV == 4
+		pim_addr mfcc_mcastgrp;
+		pim_addr mfcc_origin;
+
+		memcpy(&mfcc_mcastgrp, &c_oil->oil.mfcc_mcastgrp,
+		       sizeof(struct in_addr));
+		memcpy(&mfcc_origin, &c_oil->oil.mfcc_origin,
+		       sizeof(struct in_addr));
+
+		snprintfrr(grp_str, sizeof(grp_str), "%pPAs", &mfcc_mcastgrp);
+		snprintfrr(src_str, sizeof(src_str), "%pPAs", &mfcc_origin);
+		ifp_in = pim_if_find_by_vif_index(pim, c_oil->oil.mfcc_parent);
+#else
+		snprintfrr(grp_str, sizeof(grp_str), "%pPAs",
+			   &c_oil->oil.mf6cc_mcastgrp);
+		snprintfrr(src_str, sizeof(src_str), "%pPAs",
+			   &c_oil->oil.mf6cc_origin);
+		ifp_in = pim_if_find_by_vif_index(pim, c_oil->oil.mf6cc_parent);
+#endif
+
+		if (ifp_in)
+			strlcpy(in_ifname, ifp_in->name, sizeof(in_ifname));
+		else
+			strlcpy(in_ifname, "<iif?>", sizeof(in_ifname));
+
+		if (src_or_group) {
+			if (strcmp(src_or_group, src_str) &&
+			    strcmp(src_or_group, grp_str))
+				continue;
+
+			if (group && strcmp(group, grp_str))
+				continue;
+		}
+
+		if (json) {
+
+			/* Find the group, create it if it doesn't exist */
+			json_object_object_get_ex(json, grp_str, &json_group);
+
+			if (!json_group) {
+				json_group = json_object_new_object();
+				json_object_object_add(json, grp_str,
+						       json_group);
+			}
+
+			/* Find the source nested under the group, create it if
+			 * it doesn't exist
+			 */
+			json_object_object_get_ex(json_group, src_str,
+						  &json_source);
+
+			if (!json_source) {
+				json_source = json_object_new_object();
+				json_object_object_add(json_group, src_str,
+						       json_source);
+			}
+
+			/* Find the inbound interface nested under the source,
+			 * create it if it doesn't exist
+			 */
+			json_object_object_get_ex(json_source, in_ifname,
+						  &json_ifp_in);
+
+			if (!json_ifp_in) {
+				json_ifp_in = json_object_new_object();
+				json_object_object_add(json_source, in_ifname,
+						       json_ifp_in);
+				json_object_int_add(json_source, "Installed",
+						    c_oil->installed);
+				if (isRpt)
+					json_object_boolean_true_add(
+						json_source, "isRpt");
+				else
+					json_object_boolean_false_add(
+						json_source, "isRpt");
+				json_object_int_add(json_source, "RefCount",
```

Comment:
refCount

Reply:
probably can't change this due to preexisting use :(

(same below - all of these are just moved, they already existed before, just put into a separate file now for IPv6 support)